rfcs/0109-nixpkgs-generated-code-policy.md

```diff
@@ -0,0 +1,210 @@
+---
+feature: nixpkgs-generated-code-policy
+start-date: 2021-10-12
+author: John Ericson (@Ericson2314)
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: @L-as @sternenseemann @tomberek @DavHau
+shepherd-leader: @sternenseemann
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Nixpkgs contains non-trivial amounts of generated, rather than hand-written code.
+We want to start systematizing to make it easier to maintain.
+There is plenty of future work building upon this we could do, but we stop here for now to avoid needing to change any tools (Nix, Hydra, etc.).
+
+# Motivation
+[motivation]: #motivation
+
+Nixpkgs, along with every other distro, also faces a looming crisis: new open source software is increasingly not intended to be packaged by distros at all.
+Many languages now support very large library ecosystems, with dependencies expressed in a language-specific package manager.
+
+Right now, to deal with these packages, we either convert by hand, or commit lots of generated code into Nixpkgs.
+But I don't think either of those options is healthy or sustainable.
+The problem with the first is sheer effort; we'll never be able to keep up.
+The problem with the second is bloating Nixpkgs but more importantly reproducibility: If someone wants to update that generated code it is unclear how.
+All these mean that potential users coming from this new model of development find Nix / Nixpkgs cumbersome and unsuited to their needs.
```

Comment:
> The problem with the second is bloating Nixpkgs but more importantly reproducibility: If someone wants to update that generated code it is unclear how.

I think this needs an explanation as to why reproducibility affects simplicity of updating the packages.

I don't see a clear relationship between `reproducibility` and the `difficulty updating a package`.
We can make update mechanisms that are very easy to use without guaranteeing reproducibility, or vice versa have reproducible mechanisms which are hard to use. 

For example, an update interface like `nix run .#my-packages.update` would be very simple to use, but its result not necessarily reproducible.

Vice verse, an update mechanism can be perfectly reproducible, but if it lacks a good UI, and instead requires the user to read the source code of nixpkgs to find the right commands or attributes to trigger it, it is difficult to use despite being reproducible.

Currently the RFC seems to focus on reproducibility only. I think the statement about the UX improvement should either be dropped, or an explanation added as to how this is achieved.

> All these mean that potential users coming from this new model of development find Nix / Nixpkgs cumbersome and unsuited to their needs.

I think it makes sense to clarify which users are targeted by the RFC. Is it `nixpkgs maintainers` or `downstream users` or both?

What `downstream users` need is a way to create, update and modify package sets to their needs. For them, these interactions need to be simple, more than reproducible. The modifications become manifested in the code anyways and reproducibility is ensured from there on.
Increasing the complexity of tools in order to comply with the RFC can introduce a cost for these users, with no added benefit.

Reply:
> I think this needs an explanation as to why reproducibility affects simplicity of updating the packages.

In the sense that when it is easy to _reproduce_ the state of the generated code as present in the repository, it is easier to also effectuate desired changes. If it is, for example, not possible to update generated code without pulling in all available updates, it may be harder to just change a specific, little thing.

This would be my account of this, but perhaps the paragraph can be improved.

> Is it nixpkgs maintainers

Yes.